include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h

```diff
@@ -3,79 +3,73 @@
 
 #include <ext/intel/esimd.hpp>
 #include <cstdint>
+#include <type_traits>
+#include <limits>
 
 namespace oneapi::dpl::experimental::esimd::impl::utils
 {
 
-template<typename SIMD, typename Input>
-inline
-typename ::std::enable_if_t<::std::is_pointer_v<Input>, void>
-copy_from(SIMD& simd, const Input& input, uint32_t offset)
+template <typename T, int N>
+void
+copy_from(T* input, ::std::uint32_t base_offset, sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_from(input + offset);
+    values.copy_from(input + base_offset);
 }
 
-template<typename SIMD, typename Input>
-inline
-typename ::std::enable_if_t<!::std::is_pointer_v<Input>, void>
-copy_from(SIMD& simd, const Input& input, uint32_t offset)
+template <typename T, int N, typename... Args>
+void
+copy_from(const sycl::accessor<Args...>& input, ::std::uint32_t base_offset, sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_from(input, offset);
+    values.copy_from(input, base_offset);
 }
 
-template<typename SIMD, typename Output>
-inline
-typename ::std::enable_if_t<::std::is_pointer_v<Output>, void>
-copy_to(const SIMD& simd, Output& output, uint32_t offset)
+template <typename T, int N>
+void
+copy_to(T* output, ::std::uint32_t base_offset, const sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_to(output + offset);
+    values.copy_to(output + base_offset);
 }
 
-template<typename SIMD, typename Output>
-inline
-typename ::std::enable_if_t<!::std::is_pointer_v<Output>, void>
-copy_to(const SIMD& simd, Output& output, uint32_t offset)
+template <typename T, int N, typename... Args>
+void
+copy_to(sycl::accessor<Args...>& output, ::std::uint32_t base_offset, const sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_to(output, offset);
+    values.copy_to(output, base_offset);
 }
 
-template<typename T, int N, typename InputT>
-inline
-typename ::std::enable_if_t<::std::is_pointer_v<InputT>, sycl::ext::intel::esimd::simd<T, N>>
-gather(const InputT& input, sycl::ext::intel::esimd::simd<T, N> offsets, uint32_t base_offset)
+template <typename T, int N>
+sycl::ext::intel::esimd::simd<T, N>
+gather(T* input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets, ::std::uint32_t base_offset)
 {
-    return sycl::ext::intel::esimd::gather(input + base_offset, offsets*static_cast<uint32_t>(sizeof(T)));
+    return sycl::ext::intel::esimd::gather(input + base_offset, offsets);
```

Comment:
Computing the base address is incorrect if the offset is in bytes (which it is supposed to be).

Perhaps it makes sense to move to the element-based offset in all functions, converting to bytes when required by ESIMD API.

Reply:
Changed to use element-based offsets in the utilities. Note that the accessor variants of copy_to/from now also convert it to bytes - I believe this matches the ESIMD API, essentially fixing a bug.